### PR TITLE
kv: Don't automatically rerun replicate_queue

### PR DESCRIPTION
**Pull request by andrewbaptist**
Previously after the replicate queue completed processing a
change, it would immediately see if there were other changes
that could be done to a range. This bypassed the normal
priority order and meant that a replica might process a low
priority, expensive job before a high priority one. The intented
behavior is to check if there is additional work for the replica
and requeue at the correct priority.

Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replicate_queue.go [hunk 1]
```diff
@@ -515,37 +515,43 @@ func (rq *replicateQueue) process(
...
-			if !requeue {
-				return true, nil
+		if testingAggressiveConsistencyChecks {
+			if _, err := rq.store.consistencyQueue.process(ctx, repl, confReader); err != nil {
+				log.Warningf(ctx, "%v", err)
 			}
+		}
 
-			log.VEventf(ctx, 1, "re-processing")
+		log.VEventf(ctx, 1, "re-processing")
```
Comment (kvoli) [inline pkg/kv/kvserver/replicate_queue.go:550]: I think this line needs to go down one.

**Review by andrewbaptist — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 550 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N84-94HJFXrhOk4ZuJB-r1-550:-N84ExNoAcapztdKFLb7:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c19954a8927a2faac0576b1115690eec8b3ac0d1/pkg/kv/kvserver/replicate_queue.go#L550)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

I think this should go within the if statement below.
</blockquote></details>

Done

**Comment by andrewbaptist** [comment]
This changes the replicate_queue does not automatically run the next action, but instead queues it up to run at the appropriate priority. 

Doing this created 2 unit test failures, that I dealt with differently, but I wanted to see if anyone had concerns on this:

`TestLeasePreferencesDuringOutage` - In this case there was an expectation that after a voter was added that we would check lease preferences immediately and move the lease to the correct node if possible. This seems like a reasonable expectation, but the way it was getting there was a little convoluted. It expected a `AllocatorReplaceDeadVoter` call followed immediately by a `AllocatorConsiderRebalance` call to move the lease. With the new structure, the second call would be low priority and in a real system what would happen is that all the dead voters would be replaced before any of the rebalances happen. Adding a lease preference check after the add voter call fixed this.

`TestReplicateQueueSeesLearnerOrJointConfig` - In this case, the test is creating a situation where voters are partial complete, and it expects the replicate queue to “fix everything”. I don’t understand all the transitions in this test, especially the second one where the replicate queue has to call `AllocatorFinalizeAtomicReplicationChange` twice. I didn’t change the behavior of this test other than requiring enqueue to be called multiple times, and it might be because of the withStopAfterLearnerAtomic parameter being why the AddVoter is not successful the first time.

**Review by kvoli — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @kvoli)

___
*[`-- commits` line 6 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N86FBNH9Ps2cKA41yjS):*
nit: done to a range, rather than replica.
___
*[`-- commits` line 7 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N86FJRn7QwsQiUXgQkr):*
nit: replic -> replica.
___
*[`pkg/kv/kvserver/replicate_queue.go` line 925 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N868qNuE3kMSdDweRI2:-N868qNv0YVi3wdt5B3a:bqneuv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replicate_queue.go#L925)):*
> ```Go
> 		// We require the lease in order to process replicas, so
> 		// repl.store.StoreID() corresponds to the lease-holder's store ID.
> 		_, err := rq.shedLease(
> ```

If we did shed the lease here, then there shouldn't be a requeue for this replica because it won't have a lease. Can we set `!lhBeingRemoved=true` if `rq.ShedLease` returns true.

The replica in the queue will try and acquire the lease and (likely) fail the raft leader check its processed again, we can avoid the added processing.
___
*[`pkg/kv/kvserver/replica_learner_test.go` line 751 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N86DR6WCNw8PFc_HX9M:-N86DR6WCNw8PFc_HX9N:b-g5wvjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replica_learner_test.go#L751)):*
> ```Go
> 	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{
> 		ServerArgs:      base.TestServerArgs{Knobs: knobs},
> 		ReplicationMode: base.ReplicationManual,
> ```

since it’s in replication manual mode, the requeues in the process loop are being ignored. So it won't be testing your patch completely here.

https://github.com/cockroachdb/cockroach/blob/9b5a2b7b37b1f0648161d967ee8046d27ea96915/pkg/kv/kvserver/queue.go#L639-L641
___
*[`pkg/kv/kvserver/replica_learner_test.go` line 773 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N86ExdpAQvyrYzuNVU7:-N86ExdpAQvyrYzuNVU8:byp5ge8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replica_learner_test.go#L773)):*
> ```Go
> 		}
> 		for _, message := range expectedMessages {
> 			trace, processErr, err := store.Enqueue(
> ```

See below on manually requeuing. 
___
*[`pkg/kv/kvserver/replica_learner_test.go` line 797 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N86DGRY8BwCoK1WYaoS:-N86DGRZ8nAzifSZCNaV:b-b5eohi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replica_learner_test.go#L797)):*
> ```Go
> 			allocatorimpl.AllocatorFinalizeAtomicReplicationChange,
> 			allocatorimpl.AllocatorAddVoter,
> 			allocatorimpl.AllocatorFinalizeAtomicReplicationChange,
> ```

This is in the scope for the second stop, ltk.withStopAfterJointConfig, it will never finalize the atomic change itself until another replicate queue does it.  So there are two AllocatorFinalizeAtomicReplicationChanges here.
___
*[`pkg/kv/kvserver/replica_learner_test.go` line 801 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N86EDsp5bY_9mCCrUem:-N86EDsp5bY_9mCCrUen:bmaabqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replica_learner_test.go#L801)):*
> ```Go
> 		}
> 		for _, message := range expectedMessages {
> 			trace, processErr, err := store.Enqueue(
> ```

It would be good to test the requeuing functionality does what is expected for this test, rather than manually requeuing.

You could try disabling the replica scanner:

https://github.com/cockroachdb/cockroach/blob/9b5a2b7b37b1f0648161d967ee8046d27ea96915/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L583-L583

Then after ltk.withStopAfterLearnerAtomic you then enable the replicate queue.

Then assert on the action in processOneChange:

https://github.com/cockroachdb/cockroach/blob/9b5a2b7b37b1f0648161d967ee8046d27ea96915/pkg/kv/kvserver/replicate_queue.go#L602-L602

Likewise for the other stop.

**Review by andrewbaptist — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 925 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N868qNuE3kMSdDweRI2:-N88ycpo1CfXTejT2HgU:bz11hf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replicate_queue.go#L925)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

If we did shed the lease here, then there shouldn't be a requeue for this replica because it won't have a lease. Can we set `!lhBeingRemoved=true` if `rq.ShedLease` returns true.

The replica in the queue will try and acquire the lease and (likely) fail the raft leader check its processed again, we can avoid the added processing.


</blockquote></details>

OK, good call on this.
___
*[`pkg/kv/kvserver/replica_learner_test.go` line 751 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N86DR6WCNw8PFc_HX9M:-N89hVDA8vDsj7uVA2-R:bgxe8iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replica_learner_test.go#L751)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

since it’s in replication manual mode, the requeues in the process loop are being ignored. So it won't be testing your patch completely here.

https://github.com/cockroachdb/cockroach/blob/9b5a2b7b37b1f0648161d967ee8046d27ea96915/pkg/kv/kvserver/queue.go#L639-L641
</blockquote></details>

OK, changed the test
___
*[`pkg/kv/kvserver/replica_learner_test.go` line 773 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N86ExdpAQvyrYzuNVU7:-N89h_qe4STOP_Fl0r_D:bpx1ec7) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replica_learner_test.go#L773)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

See below on manually requeuing. 
</blockquote></details>

OK, done
___
*[`pkg/kv/kvserver/replica_learner_test.go` line 797 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N86DGRY8BwCoK1WYaoS:-N89hN864AmScENEfYU4:ba647q6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replica_learner_test.go#L797)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

This is in the scope for the second stop, ltk.withStopAfterJointConfig, it will never finalize the atomic change itself until another replicate queue does it.  So there are two AllocatorFinalizeAtomicReplicationChanges here.
</blockquote></details>

OK, makes sense
___
*[`pkg/kv/kvserver/replica_learner_test.go` line 801 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N86EDsp5bY_9mCCrUem:-N89hIVICEK7XvgAfJlP:bstec64) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replica_learner_test.go#L801)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

It would be good to test the requeuing functionality does what is expected for this test, rather than manually requeuing.

You could try disabling the replica scanner:

https://github.com/cockroachdb/cockroach/blob/9b5a2b7b37b1f0648161d967ee8046d27ea96915/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L583-L583

Then after ltk.withStopAfterLearnerAtomic you then enable the replicate queue.

Then assert on the action in processOneChange:

https://github.com/cockroachdb/cockroach/blob/9b5a2b7b37b1f0648161d967ee8046d27ea96915/pkg/kv/kvserver/replicate_queue.go#L602-L602

Likewise for the other stop.

</blockquote></details>

Done, this idea worked well

**Review by kvoli — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @andrewbaptist, and @kvoli)

___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 628 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N8PpvMp4nW5ynkKhLDK:-N8PpvMp4nW5ynkKhLDL:bov3pvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e52ca86863fcb7c56816f7c867b6663582700c81/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L628)):*
> ```Go
> 	}
> 
> 	return a.computeAction(ctx, conf, desc.Replicas().VoterDescriptors(),
> ```

Why did you drop the internal method `computeAction` here?

**Review by andrewbaptist — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @kvoli)

___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 628 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N8PpvMp4nW5ynkKhLDK:-N8Ptx-V6zwZsVHmM4X8:b-94xp7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e52ca86863fcb7c56816f7c867b6663582700c81/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L628)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Why did you drop the internal method `computeAction` here?
</blockquote></details>

The method was only called in one case by the method right above it and artificially split between what was checked in the first vs the second. I should probably not have included this change in this PR though (or maybe as a separate commit) as it is unrelated to some of the other changes. It just threw me off first when I was looking at ComputeAction vs computeAction and I was trying to understand why we needed some checks in the first method and some in the second. When I realized there was no technical reason for the split, it seemed cleaner.

**Review by kvoli — COMMENTED** [review]
This change seems significant enough to warrant some small roachprod testing to see the effects "overall" and make sure everything works as expected.

You mentioned @alextalks  has a good test. You could run the largest existing roachtest (compare to a control).

```bash
./build/builder.sh mkrelease
dev build roachtest
roachtest run decommissionBench/nodes=8/cpu=16/warehouses=3000
roachprod list
# find your cluster to get the $cluster_name - it will be your username-xxxx
roachprod adminui $cluster_name
# monitor the test.
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-:-N8Q0WP5EtJRkoYaKDeU:bgpd2vx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @andrewbaptist, and @kvoli)

___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 628 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N8PpvMp4nW5ynkKhLDK:-N8Q04FY-PuexAIPEooz:bwc2ndc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e52ca86863fcb7c56816f7c867b6663582700c81/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L628)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

The method was only called in one case by the method right above it and artificially split between what was checked in the first vs the second. I should probably not have included this change in this PR though (or maybe as a separate commit) as it is unrelated to some of the other changes. It just threw me off first when I was looking at ComputeAction vs computeAction and I was trying to understand why we needed some checks in the first method and some in the second. When I realized there was no technical reason for the split, it seemed cleaner.
</blockquote></details>

Given this pattern is used for most of the allocator methods, where there is a mostly empty exported method calling an unexported method - It would be better to leave it out of this patch, then reconsider all of them at once rather than adhoc.

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @andrewbaptist, and @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 925 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N868qNuE3kMSdDweRI2:-N8WDzpq8HBmuqdxSJmh:b-wuynd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replicate_queue.go#L925)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

OK, good call on this.
</blockquote></details>

Discussed in person - want to take a look to see if this is the correct place to do this (or what else needs to/should change).

**Review by AlexTalks — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks, @andrewbaptist, and @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 532 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N8WE8Wq2Q_1s9L5dHXD:-N8WE8Wq2Q_1s9L5dHXE:belekmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/383395815544b50b2c9de8c3687ac932fe9d98a7/pkg/kv/kvserver/replicate_queue.go#L532)):*
> ```Go
> 			// few seconds. By retrying immediately we will choose the "second best"
> 			// node to send to.
> 			// TODO: This is probably suboptimal behavior. In the case where there is
> ```

nit: `TODO(andrewbaptist)` (or, for that matter, put `TODO(sarkesian)` if needed).

I think the linter would catch this in any case.

**Review by andrewbaptist — COMMENTED** [review]
Tested with Alex, still not "ideal" behavior because stores finish at different times, but better. Alex is analyzing results

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-:-N8_FaTlAAiN-zxhwMxi:bt77ulx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @AlexTalks and @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 925 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N868qNuE3kMSdDweRI2:-N8_ICu61EtWbKnKcS3t:b-qq34fl) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d38e4689d96189d6201306669c3175d7c880388/pkg/kv/kvserver/replicate_queue.go#L925)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

Discussed in person - want to take a look to see if this is the correct place to do this (or what else needs to/should change).
</blockquote></details>

OK, The test TestLeasePreferencesDuringOutage fails without this change. Specifically, it is expecting a lease change to happen after we addOrReplaceVoters is called. This test passed before because after we re-ran the replicate queue, we would end up calling `considerRebalance` which would then rebalance and as a fortunate result also move the lease. This test was a little broken in doing what it is trying to do since if there wasn't a rebalance, this wouldn't have worked.  Now the test is a little more accurate. It might make sense long term to always call maybeTransferLeaseAway after any change is made since normally it won't do anything if lease preference is already correct, but that could be overkill. 
___
*[`pkg/kv/kvserver/replicate_queue.go` line 532 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N8WE8Wq2Q_1s9L5dHXD:-N8_FlqB3jelqTL7qu1c:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/383395815544b50b2c9de8c3687ac932fe9d98a7/pkg/kv/kvserver/replicate_queue.go#L532)):*
<details><summary><i>Previously, AlexTalks (Alex Sarkesian) wrote…</i></summary><blockquote>

nit: `TODO(andrewbaptist)` (or, for that matter, put `TODO(sarkesian)` if needed).

I think the linter would catch this in any case.
</blockquote></details>

Done
___
*[`pkg/kv/kvserver/allocator/allocatorimpl/allocator.go` line 628 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N8PpvMp4nW5ynkKhLDK:-N8TOuk4676uEGc7cxvn:bk6v7wr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e52ca86863fcb7c56816f7c867b6663582700c81/pkg/kv/kvserver/allocator/allocatorimpl/allocator.go#L628)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

Given this pattern is used for most of the allocator methods, where there is a mostly empty exported method calling an unexported method - It would be better to leave it out of this patch, then reconsider all of them at once rather than adhoc.
</blockquote></details>

Done, reverted in this PR

**Review by kvoli — APPROVED** [review]
:lgtm: - question on the TODO comment left.

> Reviewed 1 of 3 files at r3, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-:-N8_wQUk93N0l05FYyIO:bxnjack)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @AlexTalks, @andrewbaptist, and @kvoli)

___
*[`pkg/kv/kvserver/replicate_queue.go` line 1548 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85219#-N8_v_jE-7KKjJoaZcPW:-N8_v_jE-7KKjJoaZcPX:b-dwzq6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1e268073744f42a5b86a44ad43669c485e8559e/pkg/kv/kvserver/replicate_queue.go#L1548)):*
> ```Go
> // the lease and, if a suitable target is found (e.g. alive, not draining),
> // transfers the lease away.
> // TODO(baptist) change the signature of this method. The LeaseTransferOutcome
> ```

I don't think we want to remove the the error being returned, if that is what the TODO is outlining.

This error bubbles up back and is used to bump counter metrics for errors. Also when manually enqueuing, it will show up on the trace.

**Comment by andrewbaptist** [comment]
bors r=kvoli

**Pull request merged**